File: misc-runner/condenser/condenser.go
package condenser

import (
	"regexp"
	"strconv"
	"strings"

	"github.com/magic-bracket/misc-runner/types"
)

// TurnRange represents a turn segment in the log
type TurnRange struct {
	TurnNumber int
	Player     string
	StartIndex int
	EndIndex   int
}

// ExtractTurnRanges finds all turn boundaries in the log
func ExtractTurnRanges(rawLog string) []TurnRange {
	lines := strings.Split(strings.ReplaceAll(rawLog, "\r\n", "\n"), "\n")
	var ranges []TurnRange

	for i, line := range lines {
		// Try new format first: "Turn: Turn N (PlayerName)"
		if matches := ExtractTurnMarkerNew.FindStringSubmatch(line); len(matches) > 2 {
			turn, _ := strconv.Atoi(matches[1])
			ranges = append(ranges, TurnRange{
				TurnNumber: turn,
				Player:     matches[2],
				StartIndex: i,
			})
			continue
		}

		// Try old format: "Turn N: PlayerName"
		if matches := ExtractTurnMarkerOld.FindStringSubmatch(line); len(matches) > 1 {
			turn, _ := strconv.Atoi(matches[1])
			player := ""
			if len(matches) > 2 {
				player = matches[2]
			}
			ranges = append(ranges, TurnRange{
				TurnNumber: turn,
				Player:     player,
				StartIndex: i,
			})
		}
	}

	// Set end indices
	for i := range ranges {
		if i < len(ranges)-1 {
			ranges[i].EndIndex = ranges[i+1].StartIndex - 1
		} else {
			ranges[i].EndIndex = len(lines) - 1
		}
	}

	return ranges
}

// GetNumPlayers returns the number of players in the game
func GetNumPlayers(turnRanges []TurnRange) int {
	if len(turnRanges) == 0 {
		return 4 // Default for Commander
	}

	// Count unique players in turn 1
	players := make(map[string]bool)
	for _, tr := range turnRanges {
		if tr.TurnNumber == 1 && tr.Player != "" {
			players[tr.Player] = true
		}
	}

	if len(players) > 0 {
		return len(players)
	}

	return 4 // Default for Commander
}

// GetMaxRound returns the maximum round number (turn count)
func GetMaxRound(turnRanges []TurnRange, numPlayers int) int {
	if len(turnRanges) == 0 || numPlayers == 0 {
		return 0
	}

	maxTurn := 0
	for _, tr := range turnRanges {
		if tr.TurnNumber > maxTurn {
			maxTurn = tr.TurnNumber
		}
	}

	// Convert to round (full rotation)
	return (maxTurn + numPlayers - 1) / numPlayers
}

// CalculateManaPerTurn calculates mana events per round
func CalculateManaPerTurn(rawLog string, numPlayers int) map[int]types.TurnManaInfo {
	if numPlayers == 0 {
		numPlayers = 4
	}

	turnRanges := ExtractTurnRanges(rawLog)
	lines := strings.Split(strings.ReplaceAll(rawLog, "\r\n", "\n"), "\n")
	result := make(map[int]types.TurnManaInfo)

	for _, tr := range turnRanges {
		round := (tr.TurnNumber + numPlayers - 1) / numPlayers
		manaEvents := 0

		for i := tr.StartIndex; i <= tr.EndIndex && i < len(lines); i++ {
			if ExtractManaProduced.MatchString(lines[i]) || ExtractTapFor.MatchString(lines[i]) {
				manaEvents++
			}
		}

		if existing, ok := result[round]; ok {
			result[round] = types.TurnManaInfo{ManaEvents: existing.ManaEvents + manaEvents}
		} else {
			result[round] = types.TurnManaInfo{ManaEvents: manaEvents}
		}
	}

	return result
}

// CalculateCardsDrawnPerTurn calculates cards drawn per round
func CalculateCardsDrawnPerTurn(rawLog string, numPlayers int) map[int]int {
	if numPlayers == 0 {
		numPlayers = 4
	}

	turnRanges := ExtractTurnRanges(rawLog)
	lines := strings.Split(strings.ReplaceAll(rawLog, "\r\n", "\n"), "\n")
	result := make(map[int]int)

	for _, tr := range turnRanges {
		round := (tr.TurnNumber + numPlayers - 1) / numPlayers
		cardsDrawn := 0

		for i := tr.StartIndex; i <= tr.EndIndex && i < len(lines); i++ {
			line := lines[i]
			// Check for multiple draws: "draws N cards"
			if matches := ExtractDrawMultiple.FindStringSubmatch(line); len(matches) > 1 {
				if n, err := strconv.Atoi(matches[1]); err == nil {
					cardsDrawn += n
				}
			} else if ExtractDrawSingle.MatchString(line) {
				cardsDrawn++
			}
		}

		result[round] += cardsDrawn
	}

	return result
}

// ExtractWinner finds who won the game
func ExtractWinner(rawLog string) string {
	if matches := ExtractWinnerRegex.FindStringSubmatch(rawLog); len(matches) > 1 {
		return strings.TrimSpace(matches[1])
	}
	return ""
}

var ExtractWinnerRegex = regexp.MustCompile(`(?i)(.+?)\s+(?:wins\s+the\s+game|has\s+won!?)`)

// ExtractWinningTurn finds which turn the game ended on
func ExtractWinningTurn(rawLog string) int {
	lines := strings.Split(strings.ReplaceAll(rawLog, "\r\n", "\n"), "\n")
	turnRanges := ExtractTurnRanges(rawLog)
	numPlayers := GetNumPlayers(turnRanges)

	// Find the line with the win condition and determine its turn
	for i, line := range lines {
		if KeepWinCondition.MatchString(line) {
			// Find which turn range this line belongs to
			for _, tr := range turnRanges {
				if i >= tr.StartIndex && i <= tr.EndIndex {
					// Convert to round
					return (tr.TurnNumber + numPlayers - 1) / numPlayers
				}
			}
		}
	}

	// If we can't find the win line in a turn, return the last round
	if len(turnRanges) > 0 {
		return GetMaxRound(turnRanges, numPlayers)
	}

	return 0
}

// CondenseGame condenses a single raw game log into a structured summary
func CondenseGame(rawLog string) types.CondensedGame {
	// Step 1: Filter
	filteredLines := SplitAndFilter(rawLog)

	// Step 2: Classify
	keptEvents := ClassifyLines(filteredLines)

	// Step 3: Extract metrics
	turnRanges := ExtractTurnRanges(rawLog)
	numPlayers := GetNumPlayers(turnRanges)
	turnCount := GetMaxRound(turnRanges, numPlayers)
	manaPerTurn := CalculateManaPerTurn(rawLog, numPlayers)
	cardsDrawnPerTurn := CalculateCardsDrawnPerTurn(rawLog, numPlayers)

	// Step 4: Detect winner
	winner := ExtractWinner(rawLog)
	winningTurn := ExtractWinningTurn(rawLog)

	// Step 5: Build output
	condensed := types.CondensedGame{
		KeptEvents:        keptEvents,
		ManaPerTurn:       manaPerTurn,
		CardsDrawnPerTurn: cardsDrawnPerTurn,
		TurnCount:         turnCount,
	}

	if winner != "" {
		condensed.Winner = winner
	}
	if winningTurn > 0 {
		condensed.WinningTurn = winningTurn
	}

	return condensed
}

// CondenseGames condenses multiple game logs
func CondenseGames(rawLogs []string) []types.CondensedGame {
	result := make([]types.CondensedGame, len(rawLogs))
	for i, log := range rawLogs {
		result[i] = CondenseGame(log)
	}
	return result
}

// BuildAnalyzePayload builds the payload for Gemini analysis
func BuildAnalyzePayload(condensed []types.CondensedGame, deckNames []string, deckLists []string) types.AnalyzePayload {
	// Build deck info
	decks := make([]types.DeckInfo, len(deckNames))
	for i, name := range deckNames {
		decklist := ""
		if i < len(deckLists) {
			decklist = deckLists[i]
		}
		decks[i] = types.DeckInfo{
			Name:     name,
			Decklist: decklist,
		}
	}

	// Build outcomes from condensed games
	outcomes := make(map[string]types.DeckOutcome)
	for _, name := range deckNames {
		outcomes[name] = types.DeckOutcome{
			Wins:         0,
			WinningTurns: []int{},
			TurnsLostOn:  []int{},
		}
	}

	for _, game := range condensed {
		winningTurn := game.WinningTurn
		if winningTurn == 0 {
			winningTurn = game.TurnCount
		}

		for _, name := range deckNames {
			outcome := outcomes[name]
			if game.Winner != "" && strings.Contains(game.Winner, name) {
				outcome.Wins++
				outcome.WinningTurns = append(outcome.WinningTurns, winningTurn)
			} else if game.Winner != "" {
				outcome.TurnsLostOn = append(outcome.TurnsLostOn, winningTurn)
			}
			outcomes[name] = outcome
		}
	}

	return types.AnalyzePayload{
		Decks:      decks,
		TotalGames: len(condensed),
		Outcomes:   outcomes,
	}
}

File: worker/src/condenser.ts
/**
 * Log condenser - ported from misc-runner/condenser/*.go
 * Condenses raw game logs into structured summaries for AI analysis
 */

import {
  EventType,
  GameEvent,
  TurnManaInfo,
  CondensedGame,
} from './types.js';

// ============================================================================
// Regex Patterns (from patterns.go)
// ============================================================================

// Ignore patterns - lines matching these should be filtered out
const IgnorePriorityPass = /player\s+passes\s+priority/i;
const IgnoreUntapStep = /untap\s+step/i;
const IgnoreDrawStep = /draw\s+step/i;
const IgnoreBareTurn = /^Turn\s+\d+:\s*$/i;

const IgnorePatterns = [
  IgnorePriorityPass,
  IgnoreUntapStep,
  IgnoreDrawStep,
  IgnoreBareTurn,
];

// Keep patterns - lines matching these are significant events
const KeepExtraDraw = /draw(s)?\s+(an?\s+)?(additional|extra|\d+)\s+card|draw\s+\d+\s+card/i;
const KeepLifeChange = /life\s+(total\s+)?(change|loss|gain|to)|(\d+)\s+life|loses?\s+\d+\s+life|gains?\s+\d+\s+life/i;
const KeepSpellHighCMC = /cast(s|ing)?\s+.*?(?:\(?\s*CMC\s*([5-9]|\d{2,})|\(([5-9]|\d{2,})\s*\))|CMC\s*([5-9]|\d{2,})/i;
const KeepSpellCast = /\bcasts?\s+/i;
const KeepZoneChangeGYBF = /graveyard\s*->\s*battlefield|graveyard\s+to\s+battlefield|put.*from.*graveyard.*onto.*battlefield/i;
const KeepWinCondition = /wins?\s+the\s+game|game\s+over|winner|wins\s+the\s+match|loses\s+the\s+game/i;
const KeepCommanderCast = /casts?\s+(their\s+)?commander|from\s+command\s+zone/i;
const KeepCombat = /attacks?\s+with|declares?\s+attack|combat\s+damage|assigned\s+.*\s+to\s+attack/i;
const KeepLandPlayed = /^Land:/i;

// Extraction patterns
const ExtractTurnMarkerNew = /^Turn:\s*Turn\s+(\d+)\s*\((.+)\)\s*$/i;
const ExtractTurnMarkerOld = /^Turn\s+(\d+):\s*(.+?)\s*$/i;
const ExtractManaProduced = /(?:adds?|produces?|tap(s|ped)?\s+for)\s+[\w\s{}\d]*mana|(\d+)\s+mana\s+produced/i;
const ExtractTapFor = /tap(s|ped)?\s+.*?\s+for/i;
const ExtractDrawMultiple = /draws?\s+(\d+)\s+cards?/i;
const ExtractDrawSingle = /draws?\s+(?:a\s+)?card(?!s)/i;
const ExtractCMC = /\((?:CMC\s*)?(\d+)\)/i;
const ExtractWinnerRegex = /(.+?)\s+(?:wins\s+the\s+game|has\s+won!?)/i;
const GameResultPattern = /^Game Result: Game (\d+) ended/i;

// ============================================================================
// Turn Range Extraction (from condenser.go)
// ============================================================================

interface TurnRange {
  turnNumber: number;
  player: string;
  startIndex: number;
  endIndex: number;
}

function extractTurnRanges(rawLog: string): TurnRange[] {
  const lines = rawLog.replace(/\r\n/g, '\n').split('\n');
  const ranges: TurnRange[] = [];

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    // Try new format first: "Turn: Turn N (PlayerName)"
    let matches = ExtractTurnMarkerNew.exec(line);
    if (matches && matches.length > 2) {
      ranges.push({
        turnNumber: parseInt(matches[1], 10),
        player: matches[2],
        startIndex: i,
        endIndex: -1,
      });
      continue;
    }

    // Try old format: "Turn N: PlayerName"
    matches = ExtractTurnMarkerOld.exec(line);
    if (matches && matches.length > 1) {
      ranges.push({
        turnNumber: parseInt(matches[1], 10),
        player: matches[2] || '',
        startIndex: i,
        endIndex: -1,
      });
    }
  }

  // Set end indices
  for (let i = 0; i < ranges.length; i++) {
    if (i < ranges.length - 1) {
      ranges[i].endIndex = ranges[i + 1].startIndex - 1;
    } else {
      ranges[i].endIndex = lines.length - 1;
    }
  }

  return ranges;
}

function getNumPlayers(turnRanges: TurnRange[]): number {
  if (turnRanges.length === 0) {
    return 4; // Default for Commander
  }

  // Count ALL unique players across ALL turns
  const players = new Set<string>();
  for (const tr of turnRanges) {
    if (tr.player) {
      players.add(tr.player);
    }
  }

  return players.size > 0 ? players.size : 4;
}

function getMaxRound(turnRanges: TurnRange[], numPlayers: number): number {
  if (turnRanges.length === 0 || numPlayers === 0) {
    return 0;
  }

  let maxTurn = 0;
  for (const tr of turnRanges) {
    if (tr.turnNumber > maxTurn) {
      maxTurn = tr.turnNumber;
    }
  }

  // Convert to round (full rotation)
  return Math.ceil(maxTurn / numPlayers);
}

// ============================================================================
// Filter Functions (from filter.go)
// ============================================================================

function shouldIgnoreLine(line: string): boolean {
  const trimmed = line.trim();
  if (trimmed === '') {
    return true;
  }

  for (const pattern of IgnorePatterns) {
    if (pattern.test(trimmed)) {
      // Special case: Draw step with extra card draw should be kept
      if (KeepExtraDraw.test(trimmed)) {
        return false;
      }
      return true;
    }
  }

  return false;
}

function filterLines(lines: string[]): string[] {
  return lines.filter((line) => !shouldIgnoreLine(line));
}

function splitAndFilter(rawLog: string): string[] {
  const lines = rawLog.replace(/\r\n/g, '\n').split('\n');
  return filterLines(lines);
}

/**
 * Split a log that contains multiple concatenated games
 */
export function splitConcatenatedGames(rawLog: string): string[] {
  const lines = rawLog.replace(/\r\n/g, '\n').split('\n');
  const games: string[] = [];
  let currentGame: string[] = [];

  for (const line of lines) {
    if (GameResultPattern.test(line)) {
      // End of a game - save current game and start new one
      currentGame.push(line);
      games.push(currentGame.join('\n'));
      currentGame = [];
    } else {
      currentGame.push(line);
    }
  }

  // Don't forget the last game if it doesn't end with Game Result
  if (currentGame.length > 0) {
    const remaining = currentGame.join('\n').trim();
    if (remaining) {
      games.push(remaining);
    }
  }

  // If no games were split, return the original as a single game
  if (games.length === 0) {
    return [rawLog];
  }

  return games;
}

// ============================================================================
// Classification Functions (from classify.go)
// ============================================================================

function classifyLine(line: string): EventType | null {
  // Priority 1: Win Condition
  if (KeepWinCondition.test(line)) {
    return 'win_condition';
  }

  // Priority 2: Life Changes
  if (KeepLifeChange.test(line)) {
    return 'life_change';
  }

  // Priority 3: Zone Changes (Graveyard -> Battlefield)
  if (KeepZoneChangeGYBF.test(line)) {
    return 'zone_change_gy_to_bf';
  }

  // Priority 4: High CMC Spell Cast
  if (KeepSpellHighCMC.test(line)) {
    return 'spell_cast_high_cmc';
  }

  // Also check for CMC in parentheses that the main pattern might miss
  const cmcMatches = ExtractCMC.exec(line);
  if (cmcMatches && cmcMatches.length > 1) {
    const cmc = parseInt(cmcMatches[1], 10);
    if (cmc >= 5) {
      return 'spell_cast_high_cmc';
    }
  }

  // Priority 5: Commander Cast
  if (KeepCommanderCast.test(line)) {
    return 'commander_cast';
  }

  // Priority 6: Extra Card Draw
  if (KeepExtraDraw.test(line)) {
    return 'draw_extra';
  }

  // Priority 7: Combat
  if (KeepCombat.test(line)) {
    return 'combat';
  }

  // Priority 8: Land Played
  if (KeepLandPlayed.test(line)) {
    return 'land_played';
  }

  // Priority 9: Generic Spell Cast
  if (KeepSpellCast.test(line)) {
    return 'spell_cast';
  }

  return null;
}

function createEvent(
  line: string,
  turn?: number,
  player?: string
): GameEvent | null {
  const eventType = classifyLine(line);
  if (!eventType) {
    return null;
  }

  // Truncate line to 200 chars
  let truncatedLine = line.trim();
  if (truncatedLine.length > 200) {
    truncatedLine = truncatedLine.slice(0, 200);
  }

  const event: GameEvent = {
    type: eventType,
    line: truncatedLine,
  };

  if (turn !== undefined) {
    event.turn = turn;
  }
  if (player !== undefined) {
    event.player = player;
  }

  return event;
}

function classifyLines(lines: string[]): GameEvent[] {
  const events: GameEvent[] = [];
  for (const line of lines) {
    const event = createEvent(line);
    if (event) {
      events.push(event);
    }
  }
  return events;
}

// ============================================================================
// Metrics Calculation (from condenser.go)
// ============================================================================

function calculateManaPerTurn(
  rawLog: string,
  numPlayers: number
): Record<number, TurnManaInfo> {
  if (numPlayers === 0) {
    numPlayers = 4;
  }

  const turnRanges = extractTurnRanges(rawLog);
  const lines = rawLog.replace(/\r\n/g, '\n').split('\n');
  const result: Record<number, TurnManaInfo> = {};

  for (const tr of turnRanges) {
    const round = Math.ceil(tr.turnNumber / numPlayers);
    let manaEvents = 0;

    for (let i = tr.startIndex; i <= tr.endIndex && i < lines.length; i++) {
      if (ExtractManaProduced.test(lines[i]) || ExtractTapFor.test(lines[i])) {
        manaEvents++;
      }
    }

    if (result[round]) {
      result[round].manaEvents += manaEvents;
    } else {
      result[round] = { manaEvents };
    }
  }

  return result;
}

function calculateCardsDrawnPerTurn(
  rawLog: string,
  numPlayers: number
): Record<number, number> {
  if (numPlayers === 0) {
    numPlayers = 4;
  }

  const turnRanges = extractTurnRanges(rawLog);
  const lines = rawLog.replace(/\r\n/g, '\n').split('\n');
  const result: Record<number, number> = {};

  for (const tr of turnRanges) {
    const round = Math.ceil(tr.turnNumber / numPlayers);
    let cardsDrawn = 0;

    for (let i = tr.startIndex; i <= tr.endIndex && i < lines.length; i++) {
      const line = lines[i];
      // Check for multiple draws: "draws N cards"
      const multiMatches = ExtractDrawMultiple.exec(line);
      if (multiMatches && multiMatches.length > 1) {
        cardsDrawn += parseInt(multiMatches[1], 10);
      } else if (ExtractDrawSingle.test(line)) {
        cardsDrawn++;
      }
    }

    result[round] = (result[round] || 0) + cardsDrawn;
  }

  return result;
}

export function extractWinner(rawLog: string): string {
  const matches = ExtractWinnerRegex.exec(rawLog);
  if (matches && matches.length > 1) {
    return matches[1].trim().replace(/^Game outcome:\s*/i, '');
  }
  return '';
}

export function extractWinningTurn(rawLog: string): number {
  const lines = rawLog.replace(/\r\n/g, '\n').split('\n');
  const turnRanges = extractTurnRanges(rawLog);
  const numPlayers = getNumPlayers(turnRanges);

  // Find the line with the win condition and determine its turn
  for (let i = 0; i < lines.length; i++) {
    if (KeepWinCondition.test(lines[i])) {
      // Find which turn range this line belongs to
      for (const tr of turnRanges) {
        if (i >= tr.startIndex && i <= tr.endIndex) {
          // Convert to round
          return Math.ceil(tr.turnNumber / numPlayers);
        }
      }
    }
  }

  // If we can't find the win line in a turn, return the last round
  if (turnRanges.length > 0) {
    return getMaxRound(turnRanges, numPlayers);
  }

  return 0;
}

// ============================================================================
// Main Condenser Functions
// ============================================================================

/**
 * Condense a single raw game log into a structured summary
 */
export function condenseGame(rawLog: string): CondensedGame {
  // Step 1: Filter
  const filteredLines = splitAndFilter(rawLog);

  // Step 2: Classify
  const keptEvents = classifyLines(filteredLines);

  // Step 3: Extract metrics
  const turnRanges = extractTurnRanges(rawLog);
  const numPlayers = getNumPlayers(turnRanges);
  const turnCount = getMaxRound(turnRanges, numPlayers);
  const manaPerTurn = calculateManaPerTurn(rawLog, numPlayers);
  const cardsDrawnPerTurn = calculateCardsDrawnPerTurn(rawLog, numPlayers);

  // Step 4: Detect winner
  const winner = extractWinner(rawLog);
  const winningTurn = extractWinningTurn(rawLog);

  // Step 5: Build output
  const condensed: CondensedGame = {
    keptEvents,
    manaPerTurn,
    cardsDrawnPerTurn,
    turnCount,
  };

  if (winner) {
    condensed.winner = winner;
  }
  if (winningTurn > 0) {
    condensed.winningTurn = winningTurn;
  }

  return condensed;
}

/**
 * Condense multiple game logs
 */
export function condenseGames(rawLogs: string[]): CondensedGame[] {
  return rawLogs.map(condenseGame);
}

